extended validators for gossipsub v1.1 (#265)
libp2p/specs: pubsub/gossipsub/gossipsub-v1.1.md

| Lifecycle Stage | Maturity       | Status | Latest Revision |
|-----------------|----------------|--------|-----------------|

> | 1A              | Draft          | Active | r3, 2028-04-11  |

Reviewer: ```suggestion
| 1A              | Draft          | Active | r3, 2020-04-28  |
```

Not a "that far in the future" change 😅

Author: lol indeed!

`TBD`: We are currently developing multiple types of simulations that will inform us on how to best recommend tunning the Scoring function. We will update this section once that work is complete

#### Extended Validators

The pubsub subsystem incorporates application-specific message validators so that the application can
signal invalid message delivery, and trigger the P₄ penalty.

> However, It is possible to have circumstances where a message should not be delivered to the application

Reviewer 2: ```suggestion
However, it is possible to have circumstances where a message should not be delivered to the application
```

or forwarded to the network, but without triggering the P₄ penalty.

> This might arise in the case of duplicate beacon messages or while an application is syncing its blockchain, in which case it would be unable to ascertain the validity of new messages.

Reviewer: ```suggestion
A known use-case where this need exists is in the case of duplicate beacon messages or while an application is syncing its blockchain, in which case it would be unable to ascertain the validity of new messages.
```

> In order to address this situation, all gossipsub v1.1 implementations _must_ support extended validators with a ternary decision interface.

Reviewer 3: I'd prefer to keep the current cardinality out of the spec -- because it makes this a bit bespoke. I'd support making this an enum, out of which three values are supported in this version of the spec.

Author: Yes, indeed this is the intention as well, will update.

Author: done, I called it "enumerated interface".

The outcome of extended validation can be one of three things:
- Accept message; in this case the message is considered valid, and it should be delivered and forwarded to the network.
- Reject message; in this case the message is considered invalid, and it should be rejected and trigger the P₄ penalty.

> - Drop message; in this case the message is neither delivered nor forwarded to the network, but the router does not trigger the P₄ penalty.

Reviewer 3: The "reject" and "drop" terminology is confusing to me, because reject implies a drop too 🙃 

Maybe a more accurate term for the "drop" behaviour could be "ignore"?

Author: ok, I will call it `ignore`.

Author: done, called it `Ignore`.
